### PR TITLE
all: incorporate upstream linker improvements

### PR DESCRIPTION
**Pull request by josharian**
This PR pulls in the upstream linker improvements (removing relocations) and related commits. It was generated using the following script:

```zsh
#!/bin/zsh
set -e
set -v

git cherry-pick -x -Xtheirs 13aa0d8f57 # runtime: fix output for bad pcHeader
sed -i '' 's/goarch.PtrSize/sys.PtrSize/g' runtime/symtab.go
git add runtime/symtab.go
git commit --amend --no-edit -n
git cherry-pick -x 98989f2a74 # cmd/compile, cmd/link: avoid ABI aliases
git cherry-pick -x 91c2318e67 # runtime: call cgocallbackg indirectly on PPC64
git cherry-pick -x a630783ba0 # cmd/internal/obj: remove ABI aliases from object file
git cherry-pick -x 55d22f7490 # cmd/link: remove ABI aliases
git cherry-pick -x -Xtheirs e54843f2f4 # runtime: look up funcInfo by func pointer
git cherry-pick -x 6c163e5ac9 # runtime: change funcinl sentinel value from 0 to ^0
git cherry-pick -x -Xtheirs 61a0a70113 # runtime: convert _func.entry to a method
sed -i '' 's/goarch.PtrSize/sys.PtrSize/g' runtime/symtab.go
git add runtime/symtab.go
git commit --amend --no-edit -n
git cherry-pick -x f0c79caa13 # runtime: move entry method from _func to funcInfo
git cherry-pick -x 8e34d77957 # runtime, cmd/link: minor cleanup
git cherry-pick -x 45134acbe6 # debug/gosym: add pcHeader parsing helpers
git cherry-pick -x 52b23a50f8 # cmd/link: put type descriptor method arginfo in the correct section
git cherry-pick -x cd4d59232e # runtime: fix and simplify printing on bad ftab
git cherry-pick -x 850a4ffb63 # cmd/internal/goobj: remove Pcdata from object file
git cherry-pick -x 534dfb2aeb # cmd/internal/obj: refactor code to separate content-addressable symbols by section
git cherry-pick -x df63673d6a # cmd/internal/obj: index pcdata symbols in NumberSyms
git cherry-pick -x 583eeaae50 # cmd/internal/objfile: use aux symbol for pcdata references
git cherry-pick -x 587b3c1192 # cmd/link: access pcdata via aux symbols
git cherry-pick -x e180e2c27c # cmd/internal/goobj, cmd/link: remove funcdataoff
git cherry-pick -x 2fad7dbb89 # debug/gosym: add hook to disable recovers
git cherry-pick -x 10186e8d69 # debug/gosym: adjust go12* method comments
git cherry-pick -x ed57d7bb15 # debug/gosym: refactor handling of funcdata
git cherry-pick -x 88ea8a5fe0 # runtime: extract text address calculation into a separate method
git cherry-pick -x -Xtheirs d3ad216f8e # cmd/link, runtime: use offset for _func.entry
sed -i '' 's/goarch.PtrSize/sys.PtrSize/g' runtime/symtab.go
git add runtime/symtab.go
git commit --amend --no-edit -n
git cherry-pick -x 2d6d558417 # cmd/link: make funcSize a constant
git cherry-pick -x b043d231fa # debug/gosym: fix version check in file table access
git cherry-pick -x 9112d296e8 # cmd/link: add runtime.text.N symbols to Mach-O symbol table
git cherry-pick -x 96d3ba868a # cmd/link: reduce alignment for some funcdata symbols
git cherry-pick -x 8d494b0818 # cmd/link: set alignment for string symbols in symtab pass
git cherry-pick -x e5ad363fff # cmd/link: correct type..importpath symbol handling
git cherry-pick -x 579ff8b131 # cmd/go,cmd/compile: re-enable concurrent compilation on darwin/arm64
git cherry-pick -x d9952ff511 # cmd/compile: set opendefer info symbol as content-addressable
git cherry-pick -x 752cc07c77 # cmd, runtime: mark assembly routines in FuncFlags
git cherry-pick -x a8d78fae95 # cmd/compile: refactor gclocals sym creation
git cherry-pick -x -Xours c2483a5c03 # cmd, runtime: eliminate runtime.no_pointers_stackmap
goimports -w runtime/preempt.go
git add runtime/preempt.go
git commit --amend --no-edit -n
git cherry-pick -x 17674e2f17 # cmd/internal/obj, cmd/link: move symbol alignment logic to object file writer
git cherry-pick -x 199ec42350 # debug/gosym: use sort.Search in findFunc
git cherry-pick -x 78c2529d73 # debug/gosym: add funcTab abstraction
git cherry-pick -x e5f6d8d00c # cmd/internal/obj: reduce alignment for gcbits
git cherry-pick -x cfd74f7727 # cmd/link: move all FUNCDATA refs into go.func.*
git cherry-pick -x -Xtheirs 017ffcd10d # cmd/link, runtime: convert FUNCDATA relocations to offsets
git cherry-pick -x f1f626de53 # cmd/link: mangle function name with ABI on XCOFF
git cherry-pick -x -Xtheirs 3100dc1a7f # cmd/link,runtime: remove relocations from stkobjs
git cherry-pick -x c1a0aa300a # cmd/compile: make stkobj symbols content-addressable
git cherry-pick -x 113b52979f # runtime: remove a branch from funcdata
git cherry-pick -x e31c9ab557 # cmd/link,runtime: remove functab relocations
git cherry-pick -x -Xtheirs 77bd0da688 # cmd/link,runtime: remove unnecessary funcdata alignment
sed -i '' 's/goarch.PtrSize/sys.PtrSize/g' runtime/symtab.go
git add runtime/symtab.go
git commit --amend --no-edit -n
git cherry-pick -x 5758c40ac8 # runtime: add a single-text-section fast path to findfunc
git cherry-pick -x -Xtheirs 2a5d4ea97e # runtime: make funcspdelta inlineable
sed -i '' 's/goarch.PtrSize/sys.PtrSize/g' runtime/symtab.go
git add runtime/symtab.go
git commit --amend --no-edit -n
git cherry-pick -x e82ed0cd83 # runtime: start moduledata memory load early
git cherry-pick -x 415f0a3375 # cmd/link: keep go.buildid alive on AIX external linking
git cherry-pick -x 0e13da8fae # cmd/internal/obj: allow more gcbits content addressability
git cherry-pick -x 6487b1573e # cmd/link: remove unnecessary int conversions
git cherry-pick -x 058fa255bc # cmd/link,runtime: make textsectmap fields more convenient for runtime
git cherry-pick -x 8238f82bf1 # runtime: streamline moduledata.textAddr
./all.bash
```

### CODE REVIEW

**Comment by josharian** [comment]
This is not ready to merge yet. It passes all.bash locally, but that is all the testing that I have done.

**Comment by bradfitz** [comment]
Well, at least it was a clean cherry pick! 😜

**Comment by bradfitz** [comment]
Glad you kept AIX working in our tree!

**Comment by josharian** [comment]
> Glad you kept AIX working in our tree!

Shhhhhhhh, don't spoil the big Tailscale 2.0 reveal. :P

**Comment by crawshaw** [comment]
The diff is of course, impossible to read. :) But your script seems quite safe, with the caveat that I've never messed with a merge strategy and have no idea what -Xtheirs does. LGTM once all those cherry-pick sources are upstream.

**Comment by steeve** [comment]
outstanding work. can those be applied on top of 1.17.x ?

File: src/cmd/internal/obj/objfile.go [hunk 1]
```diff
@@ -397,7 +393,39 @@ func (w *writer) Hash(s *LSym) {
...
+// The goal is to prevent content-addressability from moving symbols between sections.
+// contentHashSection only distinguishes between sets of sections for which this matters.
+// Allowing flexibility increases the effectiveness of content-addressibility.
+// But in some cases, such as doing addressing based on a base symbol,
+// we need to ensure that a symbol is always in a prticular section.
+// Some of these conditions are duplicated in cmd/link/internal/ld.(*Link).symtab.
+// TODO: instead of duplicating them, have the compiler decide where symbols go.
+func contentHashSection(s *LSym) byte {
+	name := s.Name
+	if s.IsPcdata() {
```
Comment (jcchavezs) [inline src/cmd/internal/obj/objfile.go:406]: nit: do you still want to allocate `name` when next line is an early return that does not use it?

**Review by bradfitz — CHANGES_REQUESTED** [review]
GitHub bot pinged me about this, so I'm replying.

I think we decided to hold on this until Go 1.18?

**Comment by josharian** [comment]
Yep. Closing this. It'll come along for the ride with the rest of Go 1.18.

**Comment by josharian** [comment]
(not deleting the branch, though, just in case)

**Pull request closed without merging**